skimage/filters/edges.py

```diff
@@ -55,65 +57,158 @@ def _mask_filter_result(result, mask):
+        The requested shape.
+
+    Examples
+    --------
+    >>> kernel_shape(2, 0)
+    [-1, 1]
+    >>> kernel_shape(3, 1)
+    [1, -1, 1]
+    >>> kernel_shape(4, -1)
+    [1, 1, 1, -1]
+    """
+    shape = [1, ] * ndim
+    shape[dim] = -1
+    return shape
+
+
+def _generic_edge_filter(image, *, smooth_weights, edge_weights=[1, 0, -1],
+                         axis=None, mode='reflect', cval=0.0, mask=None):
+    """Apply a generic, n-dimensional edge filter.
+
+    The filter is computed by applying the edge weights along one dimension
+    and the smoothing weights along all other dimensions. If no axis is given,
+    or a tuple of axes is given the filter is computed along all axes in turn,
+    and the magnitude is computed as the square root of the average square
+    magnitude of all the axes.
+
+    Parameters
+    ----------
+    image : array
+        The input image.
+    smooth_weights : array of float
+        The smoothing weights for the filter. These are applied to dimensions
+        orthogonal to the edge axis.
+    edge_weights : 1D array of float, optional
+        The weights to compute the edge along the chosen axes.
+    axis : int or sequence of int, optional
+        Compute the edge filter along this axis. If not provided, the edge
+        magnitude is computed. This is defined as::
+
+            edge_mag = np.sqrt(sum([_generic_edge_filter(image, ..., axis=i)**2
+                                    for i in range(image.ndim)]) / image.ndim)
+
+        The magnitude is also computed if axis is a sequence.
+    mode : str or sequence of str, optional
+        The boundary mode for the convolution. See `scipy.ndimage.convolve`
+        for a description of the modes. This can be either a single boundary
+        mode or one boundary mode per axis.
+    cval : float, optional
+        When `mode` is ``'constant'``, this is the constant used in values
+        outside the boundary of the image data.
+    """
+    ndim = image.ndim
+    if axis is None:
+        axes = list(range(ndim))
+    elif np.isscalar(axis):
+        axes = [axis]
     else:
-        mask = binary_erosion(mask, EROSION_SELEM, border_value=0)
-        return result * mask
+        axes = axis
+    return_magnitude = (len(axes) > 1)
+
+    output = np.zeros(image.shape, dtype=float)
 
+    for edge_dim in axes:
+        kernel = np.reshape(edge_weights, kernel_shape(ndim, edge_dim))
+        smooth_axes = list(set(range(ndim)) - {edge_dim})
+        for smooth_dim in smooth_axes:
+            kernel = kernel * np.reshape(smooth_weights,
+                                         kernel_shape(ndim, smooth_dim))
+        ax_output = ndi.convolve(image, kernel, mode='reflect')
+        if return_magnitude:
+            ax_output *= ax_output
+        output += ax_output
 
-def sobel(image, mask=None):
-    """Find the edge magnitude using the Sobel transform.
+    if return_magnitude:
+        output = np.sqrt(output) / np.sqrt(ndim)
```

Comment:
It's not a bad plan. How about `output = np.sqrt(output, out=output)`, since the array is returned anyway? To me that makes more readable, while maintaining the performance benefits. Anyway, I'll see whether the difference in performance is worth it.

Reply:
@grlee77 I see about a 10% speedup in 2D, and 0% in 3D, in exchange for a bit of numerical instability (some of our tests, which admittedly might be poorly designed, start failing with this new form.), and less readable code. So I think I'd rather leave this as is.

Reply:
Thanks for checking.